crates/red_knot_python_semantic/resources/mdtest/narrow/isinstance.md

```diff
@@ -0,0 +1,63 @@
+# Narrowing for `isinstance` checks
+
+Narrowing for `isinstance(object, classinfo)` expressions.
+
+## `classinfo` is a single type
+
+```py
+x = 1 if flag else "a"
+
+if isinstance(x, int):
+    reveal_type(x)  # revealed: Literal[1]
+
+if isinstance(x, str):
+    reveal_type(x)  # revealed: Literal["a"]
+    if isinstance(x, int):
+        reveal_type(x)  # revealed: Never
+
+if isinstance(x, (int, object)):
+    reveal_type(x)  # revealed: Literal[1] | Literal["a"]
+```
+
+## `classinfo` is a tuple of types
+
+Note: `isinstance(x, (int, str))` should not be confused with
+`isinstance(x, tuple[(int, str)])`. The former is equivalent to
+`isinstance(x, int | str)`:
```

Comment:
I can't add a test for `isinstance(…, tuple[(int, str)])` because we don't support generics yet.

Reply:
`isinstance(..., tuple[(int, str)])` doesn't work at runtime:

```
>>> x = (1, "a")
>>> isinstance(x, tuple)
True
>>> isinstance(x, tuple[(int, str)])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: isinstance() argument 2 cannot be a parameterized generic
```

So we should (eventually, not in this PR) emit a diagnostic for this in type inference, and we don't need to support it in narrowing.

(Also, FWIW, although we don't support generalized generics yet, we do have a representation for heterogeneous tuple types specifically, since they are a special case, and it wouldn't be very hard to add support in type-expression inference for understanding `tuple[int, str]` as a type annotation, if that were blocking something else.)

Reply:
I _think_ we should also already accurately infer the _runtime_ value of `tuple[int, str]` as being a `types.GenericAlias` instance:

```pycon
>>> tuple[int, str]
tuple[int, str]
>>> type(_)
<class 'types.GenericAlias'>
```